Standards (#24)
hackoregon/2019-backend-cookiecutter-django: docs/standards/v1/STANDARDS.md

We aim to also provide the highest level of interoperability between our APIs. By adopting standards between teams, we can better support users connecting to multiple APIs.

### API Technologies
* **HTTPS:** All API requests must be made over HTTPS.
* **JSON:** All API request and response bodies will be JSON.

## Requests

Requests and endpoints should follow a common REST-based structure.

### URL Structure

* URLs will use the Hack Oregon team name as the root:  
  Pattern: `:team_name/v1/:resource name`   
  Example: `transportation_systems/v1/stops`  

* URLs will be prefixed by the major version number:  
  Pattern: `/v1/:resource name`  
  Example: `/v1/stops`  

* Collections of resources are referenced by their resource name (plural):  
  Pattern: `/v1/:resource name`  
  Example: `/v1/stops`

* Individual resources are referenced by their resource name (plural) followed by the guid:  
  Pattern: `/v1/:resource name/:guid`  
  Example: `/v1/stops/25fe21b8-8de2-40d0-93b0-c819101d1a11`

### GET

Request a single resource or collection of Resources

* GET requests may include query parameters

> * GET requests must NOT include a request body

Author: https://github.com/hackoregon/2019-backend-cookiecutter-django/commit/00bc76bdd80333988672fef3f9ff425dd0d86058

#### Single Resource

```
GET /v1/stops/25fe21b8-8de2-40d0-93b0-c819101d1a11
```

##### Responses (Resource)

|Scenario|Code|Body|
|---|---|---|
| Valid Resource/Request | 200 | Resource |
| Bad Request | 400 | Error |
| Not Found | 404 | Error |

> | Internal Server Error | 500 | Error |

Author: https://github.com/hackoregon/2019-backend-cookiecutter-django/commit/00bc76bdd80333988672fef3f9ff425dd0d86058

#### Collection

```
GET /v1/stops/
```

##### Responses (Resource)

|Scenario|Code|Body|
|---|---|---|
| Valid Resource/Request | 200 | List of Resources (All or Paginated) |

> | Valid Resource, No Results | 200 | Empty Body|

Author: https://github.com/hackoregon/2019-backend-cookiecutter-django/commit/00bc76bdd80333988672fef3f9ff425dd0d86058

> | Bad Request | 400 | Error |

Author: https://github.com/hackoregon/2019-backend-cookiecutter-django/commit/00bc76bdd80333988672fef3f9ff425dd0d86058

| Internal Server Error | 500 | Error |

#### Recommended Technologies to Support

> * Django Restframework Viewsets and Mixins should generally handle this behavior. If you are writing custom views, use built-in status codes.

Reviewer: Add links to Viewsets and Mixins documentation
```suggestion
* Django Rest Framework [Viewsets](https://www.django-rest-framework.org/api-guide/viewsets/) and [Mixins](https://www.django-rest-framework.org/api-guide/generic-views/#mixins) should generally handle this behavior. If you are writing custom views, use built-in status codes.
```

> * Users will have to correctly provide routing

Reviewer: Specify which users (API developers? API consumers?)

Author: clarified:

"* API developers will have to correctly define routing"

https://github.com/hackoregon/2019-backend-cookiecutter-django/commit/45f51e02c73641704f2249d58bd21f813d56324c

> * Django Rest Framework allows versioning to be handled [versioning](https://www.django-rest-framework.org/api-guide/versioning/) through a few methods.

Reviewer: ```suggestion
* Django Rest Framework allows [versioning](https://www.django-rest-framework.org/api-guide/versioning/) to be handled through a few methods.
```

## Response Codes

### Successful Requests

|Status Code|Description|Verbs|
|---|---|---|
|200 OK|This status **MUST** be returned for synchronous requests that complete successfully and have a response body. This **MUST** only be used if there is not a more appropriate 2XX response code. |GET, PATCH|

### Client Errors

|Status Code|Description|Verbs|
|---|---|---|

> |400 Bad Request|This status **MUST** be returned for requests that provide malformed or invalid data. Examples: malformed request body, unexpected query parameters, or invalid request fields.|GET, PATCH, POST, DELETE|

Author: https://github.com/hackoregon/2019-backend-cookiecutter-django/commit/ca51d2289feb531ea82fc1e7a0f022b522e18d75

|403 Forbidden|This status **MUST** be returned if the request cannot be performed by the user due to lack of permissions. Example: User with read-only permissions to a resource tries to update it. |POST, PATCH, DELETE|
|404 Not Found|This status **MUST** be returned if the requested resource does not exist or if the user requesting the resource has insufficient permissions to view the resource.|GET, POST, PATCH, DELETE|

### Server Errors

|Status Code|Description|
|---|---|
|500 Internal Server Error|This status **MUST** be returned when a server error occurs.

## Resources

A resource represents an individual object within the system.  It is represented as a JSON object.  

## Field Names

Resource Fields will be camelCase.

Resource Fields MUST include ONLY the following characters:

* `a-z` (lowercase alpha)
* `A-Z` (uppercase alpha)

Resource fields that accept multiple values MUST be pluralized.

#### Recommended Technologies to Support

* https://github.com/vbabiy/djangorestframework-camel-case

## Collections

> A collection is a list of multiple Resources. A collection is represented as a JSON object.

Author: will add

> A collection is a list of multiple Resources. A collection is represented as a JSON object.

Author: Added: https://github.com/hackoregon/2019-backend-cookiecutter-django/commit/f4047cccab1d9356e8d5f233528790344177998c

If a collection is larger then per_page Pagination size then it `MUST` contain a links object, containing pagination information.

## Pagination

Pagination should be used by Collections to limit the number of resources returned at a time.

Pagination is requested by a client through the use of query parameters.

Requests will accept the following query parameters:

- limit: maximum number of items to return
- offset: starting position of the query in relation to the complete set of unpaginated items

A paginated response would look as follows:

```json
{
    "count": 1023,
    "next": "https://api.example.org/accounts/?limit=100&offset=500",
    "previous": "https://api.example.org/accounts/?limit=100&offset=300",
    "results": [
       …
    ]
}
```
#### Recommended Technologies to Support

To achieve this, register a `DEFAULT_PAGINATION_CLASS` in your `local_settings/settings.py` to apply to all apps:

```python
REST_FRAMEWORK = {
    "DEFAULT_PAGINATION_CLASS": "rest_framework.pagination.LimitOffsetPagination",
    "PAGE_SIZE": 10
}
```

> PAGE_SIZE should be set to:  "TO BE DETERMINED"

Reviewer 2: I think it's going to be hard to come up with a one-size-fits-all default given that it will vary on resource size and query complexity, but I certainly think something larger than 10 should be the recommendation 😄

Author: https://github.com/hackoregon/2019-backend-cookiecutter-django/commit/37aa17e08fb80bad6deee2fe21e39db2a4419e55

## Query Parameters

Query Parameters **MUST** include **ONLY** the following characters:

* `a-z` (lowercase only)
* `_` (underscore)

Query parameters that accept multiple values **MUST** be pluralized.

> If any request receives a query parameter it does not understand, the response **MUST** be a `400 Bad Request`.

Reviewer 2: This is an interesting one. 

I see the benefit of being strict here, but I can see it also being a pain from an API consumer standpoint. Sometimes it's nice to serialize an entire object as query params without having to cherry-pick out any values that aren't valid.

This is how GitHub's API works: https://api.github.com/orgs/hackoregon/repos?sort=updated&beepboop=boopbeep!

It's also inline with the [robustness principle](https://en.wikipedia.org/wiki/Robustness_principle).

Meanwhile the [JSON:API spec](https://jsonapi.org/format/#query-parameters) doesn't take a stance on query parameters as long as there is no chance they could conflict with params that may be added to a future spec.

The [OpenAPI spec](http://spec.openapis.org/oas/v3.0.2) also doesn't seem concerned about extraneous query params.

Author: Removed requirement:

https://github.com/hackoregon/2019-backend-cookiecutter-django/commit/6d48f23f5e2422f87decaafd4475954f845ebb70

                         [101.0, 1.0],
                         [100.0, 1.0],
                         [100.0, 0.0]
                     ]
                 ]
             },
             "properties": {
                 "firstProp": "value0",
                 "secondProp": {
                     "this": "that"
                 }
             }
         }]
     }
  ```
  * Standard Projection should be [EPSG:4326](https://spatialreference.org/ref/epsg/wgs-84/)

#### Recommended Technologies to Support

* [djangorestframework-gis](https://github.com/djangonauts/django-rest-framework-gis) provides support for DRF for models, views, serializers, returning valid GeoJSON
* Use with [POSTGIS](https://postgis.net/) database and [Django PostGIS engine](https://docs.djangoproject.com/en/2.2/ref/contrib/gis/tutorial/#configure-settings-py)

## Documentation

Projects should include the following documentation:

* Basic README providing project summary, Hack Oregon intro, quickstart, installation, and contribution guidelines
* Formatted [docstrings](https://www.django-rest-framework.org/api-guide/schemas/#schemas-as-documentation) on classes and methods
* Swagger UI frontend

> * Published OpenAPI schema

Reviewer: > Swagger UI frontend
> Published OpenAPI schema

Would API developers need any additional links or information in this standards document about how to implement these items?

Author: Yes we can provided further details, though I do want to get the general guide approved, and moved to master for visibility.

I added this note to spec, which I think we can update in the next 1-2 weeks, once some of the documentation open issues are complete?

https://github.com/hackoregon/2019-backend-cookiecutter-django/commit/e5deb44f4e905c7f76147c1c0d963049d3df8364

# Hack Oregon Data APIs v1 Style Guide

## Overview

The intent of this document is to provide some guidance for API developers when building APIs under Hack Oregon Projects. Our aim is to follow general RESTful API best practices, and work with solutions/standards that are built into our technology stack.

We aim to also provide the highest level of interoperability between our APIs. By adopting standards between teams, we can better support users connecting to multiple APIs.

### API Technologies
* **HTTPS:** All API requests must be made over HTTPS.

> * **JSON:** All API request and response bodies will be JSON.

Reviewer 3: From the examples, it seems we're following JSON:API. Can we state that explicitly and link to [https://jsonapi.org/format/](https://jsonapi.org/format/)?

      "value": "8609518.76667",
      "rank": 1,
      "total": 5
    }
  ]
}

```

## Pagination

Pagination **SHOULD** be used by Collections to limit the number of resources returned at a time.

Pagination is requested by a client through the use of query parameters.

Requests will accept the following query parameters:

- limit: maximum number of items to return
- offset: starting position of the query in relation to the complete set of unpaginated items

A paginated response would look as follows:

```json
{
    "count": 1023,
    "next": "https://api.example.org/accounts/?limit=100&offset=500",
    "previous": "https://api.example.org/accounts/?limit=100&offset=300",
    "results": [
       …
    ]
}
```
#### Recommended Supporting Technologies

To achieve this, register a `DEFAULT_PAGINATION_CLASS` in your `local_settings/settings.py` to apply to all apps:

```python
REST_FRAMEWORK = {
    "DEFAULT_PAGINATION_CLASS": "rest_framework.pagination.LimitOffsetPagination",
    "PAGE_SIZE": 50
}
```

PAGE_SIZE is recommended to be between 50-100, depending on the size of resource object.

## Query Parameters

Query Parameters **MUST** include **ONLY** the following characters:

> * `a-z` (lowercase only)

Reviewer: should also allow 0-9?

* `_` (underscore)

Query parameters that accept multiple values **MUST** be pluralized.

All query parameters **MUST** be properly [url-encoded](https://en.wikipedia.org/wiki/Percent-encoding#Current_standard). If a single query parameter value includes the comma (`,`) character, the comma **MUST** be double encoded.

> **Note:** For readability purposes, the examples throughout this document do not show encoded query strings.

#### Examples
Single value:
`GET /v1/apps?names=firstname`

Multiple values:
 `GET /v1/apps?names=firstname,secondname`

> Single value with comma:

Reviewer: ```suggestion
Multiple values with double encoded commas:
```

>  `GET /v1/apps?names=comma%2Cname`

Reviewer: ```suggestion
 `GET /v1/apps?address=Portland%2cOR,Vancouver%2cWA `
```

## Filtering

Filtering is the use of query parameters to return a subset of resources within a [Collection](#collections).

Filter query parameters **MUST** have names that conform to the acceptable [query parameter](#query-parameters) names.
Filter parameters **MUST** be able to be combined with other filters on the same collection.

When multiple filters are provided, the results **MUST** match all specified filters.

* All filters, including pagination filters should be available to the schema autogeneration

#### Examples

**Single value request**:
`GET /v1/apps?names=the_name`

This will return all apps with name `the_name`.

**Multiple value request**:
`GET /v1/apps?names=first_name,second_name`

This will return all apps with name `the_name` OR `second_name`.

**Combined filters**:
`GET /v1/apps?names=the_name&state=STARTED`

This will return all apps with name `the_name` AND state `STARTED`.

#### Recommended Supporting Technologies

> * [django-filter](https://www.django-rest-framework.org/api-guide/filtering/#djangofilterbackend) and the DjangoFilterBackend is encouraged to be used.

Reviewer: ```suggestion
* [django-filter](https://www.django-rest-framework.org/api-guide/filtering/#djangofilterbackend) implementing the [DjangoFilterBackend](https://www.django-rest-framework.org/api-guide/filtering/#djangofilterbackend) is encouraged to be used.
```

> * Create a [`get_schema_fields(self, view)`](https://www.django-rest-framework.org/api-guide/filtering/#pagination-schemas) method on a view in which you need to add fields to schema autogeneration (uses [CoreAPI](https://www.django-rest-framework.org/api-guide/schemas/#core-api)). May need to do this to add [pagination](#pagination) filters.

Reviewer: ```suggestion
* Create a [`get_schema_fields(self, view)`](https://www.django-rest-framework.org/api-guide/filtering/#pagination-schemas) method on a view to add fields to schema autogeneration (uses [CoreAPI](https://www.django-rest-framework.org/api-guide/schemas/#core-api)). May need to do this to add [pagination](#pagination) filters.
```
